crates/matrix-sdk-ui/src/timeline/inner/mod.rs

```diff
@@ -434,41 +442,41 @@ impl<P: RoomDataProvider> TimelineInner<P> {
 
         // The local echoes are always at the end of the timeline, we must first make
         // sure the remote echo hasn't showed up yet.
-        if rfind_event_item(&state.items, |it| {
+        if rfind_event_item(&items_txn, |it| {
             new_event_id.is_some() && it.event_id() == new_event_id && it.as_remote().is_some()
         })
         .is_some()
         {
             // Remote echo already received. This is very unlikely.
             trace!("Remote echo received before send-event response");
 
-            let local_echo =
-                rfind_event_item(&state.items, |it| it.transaction_id() == Some(txn_id));
+            let local_echo = rfind_event_item(&items_txn, |it| it.transaction_id() == Some(txn_id));
 
             // If there's both the remote echo and a local echo, that means the
             // remote echo was received before the response *and* contained no
             // transaction ID (and thus duplicated the local echo).
             if let Some((idx, _)) = local_echo {
                 warn!("Message echo got duplicated, removing the local one");
-                state.items.remove(idx);
+                items_txn.remove(idx);
 
                 if idx == 0 {
                     error!("Inconsistent state: Local echo was not preceded by day divider");
                     return;
                 }
 
-                if idx == state.items.len() && state.items[idx - 1].is_day_divider() {
+                if idx == items_txn.len() && items_txn[idx - 1].is_day_divider() {
                     // The day divider may have been added for this local echo, remove it and let
                     // the next message decide whether it's required or not.
-                    state.items.remove(idx - 1);
+                    items_txn.remove(idx - 1);
                 }
             }
 
+            items_txn.commit();
```

Comment:
Does the `commit()` not always happen on Drop anyways? Since we return on the next line, that will drop the transaction, right?

Reply:
Commit on drop is removed in the last git commit (it's only introduced to have a smaller diff for initial txn usage). That's part of why I wasn't sure whether the last two commits should be separate at all.